### PR TITLE
support obkv rpc buffer compress

### PR DESCRIPTION
**Pull request by GroundWu**
## Summary

support obkv rpc buffer compress

### CODE REVIEW

File: obkvrpc/connection.go [hunk 1]
```diff
@@ -481,6 +481,21 @@ func (c *Connection) decodePacket(contentBuf []byte, response protocol.ObPayload
...
+		decompressor, err := getDecompressor(rpcHeader.CompressType())
+		if err != nil {
+			return errors.New("fail to get decompressor")
+		}
+		decompressBuffer, err := decompressor.Decompress(contentBuffer, rpcHeader.OriginalLen())
+		if err != nil {
+			return err
+		}
+		contentBuffer = decompressBuffer
+		log.Info(fmt.Sprintf("compressType: %s, compressLen: %d, originLen: %d\n",
```
Comment (WeiXinChan) [inline obkvrpc/connection.go]: too more log

File: obkvrpc/compressor.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	// 1. create zlib reader
+	zlibReader, err := zlib.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func(zlibReader io.ReadCloser) {
+		err := zlibReader.Close()
+		if err != nil {
+			panic(err.Error())
```
Comment (WeiXinChan) [inline obkvrpc/compressor.go:81]: print log_warn, should not panic

File: obkvrpc/compressor.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+func (d *Lz4Decompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes := make([]byte, originLen)
+	n, err := lz4.UncompressBlock(srcBytes, dstBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
```
Comment (WeiXinChan) [inline obkvrpc/compressor.go:54]: Lz4Decompressor: length of decompress content...

File: obkvrpc/compressor.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+}
+
+func (d *SnappyDecompressor) Decompress(src *bytes.Buffer, originLen int32) (*bytes.Buffer, error) {
+	srcBytes := src.Bytes()
+	dstBytes, err := snappy.Decode(nil, srcBytes)
+	if err != nil {
+		return nil, err
+	}
+	if int32(len(dstBytes)) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", int32(len(dstBytes)), originLen))
```
Comment (WeiXinChan) [inline obkvrpc/compressor.go:66]: SnappyDecompressor: length of decompress content

File: obkvrpc/compressor.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+	// 2. create buffer
+	dstBuffer := new(bytes.Buffer)
+	// 3. write the decompress data to dstBuffer
+	n, err := io.Copy(dstBuffer, zlibReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
```
Comment (WeiXinChan) [inline obkvrpc/compressor.go:93]: ZlibDecompressor: length of decompress content i...

File: obkvrpc/compressor.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		return nil, err
+	}
+	defer zstdReader.Close()
+	dstBuffer := new(bytes.Buffer)
+	n, err := io.Copy(dstBuffer, zstdReader)
+	if err != nil {
+		return nil, err
+	}
+	if int32(n) != originLen {
+		return nil, errors.New(fmt.Sprintf("length of decompress content is not expected: actual-%d, expected-%d", n, originLen))
```
Comment (WeiXinChan) [inline obkvrpc/compressor.go:110]: ZStdDecompressor: length of decompress content ...

File: obkvrpc/connection.go (hunk 1 above)
Comment (GroundWu) [reply obkvrpc/connection.go]: use log.Debug instead

File: obkvrpc/compressor.go (hunk 2 above)
Comment (GroundWu) [reply obkvrpc/compressor.go]: done

File: obkvrpc/compressor.go (hunk 3 above)
Comment (GroundWu) [reply obkvrpc/compressor.go]: done

File: obkvrpc/compressor.go (hunk 4 above)
Comment (GroundWu) [reply obkvrpc/compressor.go]: done

File: obkvrpc/compressor.go (hunk 5 above)
Comment (GroundWu) [reply obkvrpc/compressor.go]: done

File: obkvrpc/compressor.go (hunk 6 above)
Comment (GroundWu) [reply obkvrpc/compressor.go]: done

File: obkvrpc/compressor.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	case protocol.ObCompressTypeSnappy:
+		return "snappy"
+	case protocol.ObCompressTypeZlib:
+		return "zlib"
+	case protocol.ObCompressTypeZstd:
+		return "zStd"
+	case protocol.ObCompressTypeNone:
+		return "none"
+	case protocol.ObCompressTypeZstd_138:
+		return "zstd_1.3.8"
```
Comment (IHEII) [inline obkvrpc/compressor.go:150]: It seems different from 
``` source
    return "zStd"
```
Keep _zstd_ written in the same way

File: obkvrpc/compressor.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	case protocol.ObCompressTypeLZ4:
+		return &Lz4Decompressor{}, nil
+	case protocol.ObCompressTypeSnappy:
+		return &SnappyDecompressor{}, nil
+	case protocol.ObCompressTypeZlib:
+		return &ZlibDecompressor{}, nil
+	case protocol.ObCompressTypeZstd:
+		return &ZStdDecompressor{}, nil
+	case protocol.ObCompressTypeZstd_138:
+		return &ZStdDecompressor{}, nil
```
Comment (IHEII) [inline obkvrpc/compressor.go:127]: Will ObCompressTypeZstd_138 and ObCompressTypeZstd use the same decompressor and be processed the same way?

File: test/compress_rpc_result/all_test.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	cli = test.CreateClient()
+	test.CreateDB()
+	test.CreateTable(createTableStat)
+}
+
+func setCompressType(compressTyepe string) error {
+	if test.GlobalDB == nil {
+		return errors.New("test GlobalDb is nil")
+	}
+	CompressFuncSql := "alter system set kv_transport_compress_func= '%s'"
```
Comment (IHEII) [inline test/compress_rpc_result/all_test.go:42]: Make this SQL a  const value outside.

File: obkvrpc/compressor.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (GroundWu) [reply obkvrpc/compressor.go]: it will print warn log instead

File: test/compress_rpc_result/all_test.go (hunk 9 above)
Comment (GroundWu) [reply test/compress_rpc_result/all_test.go]: done

File: obkvrpc/compressor.go (hunk 8 above)
Comment (GroundWu) [reply obkvrpc/compressor.go:127]: Yes. These two compress types are supported by observer rpc framework. They are the two different version of the same compress algorithms. It has been verified that can use the same decompressor to process in client and test case has been added in `query_with_compress_test.go`.

File: obkvrpc/compressor.go (hunk 7 above)
Comment (GroundWu) [reply obkvrpc/compressor.go:150]: I think it return the "zstd_1.3.8" and "zstd" may be a better way to classify which compress func observer used

**Comment by WeiXinChan** [comment]
LGTM

File: obkvrpc/compressor.go (hunk 7 above)
Comment (IHEII) [reply obkvrpc/compressor.go:150]: We should use _zStd_1.3.8_ in line 150 or _zstd_  in line 146.
Keep "s" in the same format.

**Comment by shenyunlong** [comment]
LGTM

File: obkvrpc/compressor.go (hunk 7 above)
Comment (GroundWu) [reply obkvrpc/compressor.go:150]: done, use 'zstd' to keep the same format

**Pull request merged**